### PR TITLE
Implement basic Auth0

### LINKED ISSUE #279: Implement user authentication
Once we ship #278, we'll need some way for users to sign up and log in. Here are some options I've considered so far:

### External provider

We could use Netlify Identify, Auth0, Firebase, or something similar.

Pros: very fast/easy
Cons: vendor lock-in, possibly limited by cost in the future, lack of flexibility, giving user data to another company

### Password auth

Email/password is the most obvious. We'd end up needing to implement some basic functionality like password resets, email confirmations. There may already be a Node package for doing this?

Pros: straight forward, simple
Cons: need to implement support for typical account management features

### Slack login

What if started with "Log in with Slack" and restricted accounts to people on any climate Slack groups: MCJ, CAT, Impact Makers, Circular Carbon, etc...

Pros: no user account management functionality needed, naturally establishes a limited "beta test" group for launch
Cons: non-Slack users are left out

### PR DESCRIPTION
**Pull request by bloudermilk**
Work in progress, adding Auth0 to the site. Fixes #279

### CODE REVIEW

File: site/gatsby-browser.js [hunk 1]
```diff
@@ -4,4 +4,17 @@
...
+
+exports.wrapRootElement = ({ element }) => (
+  <Auth0Provider
+    cacheLocation="localstorage"
+    redirect_uri={location.origin + "/authback"}
+    onRedirectCallback={() => navigate("/")}
+  >
+    {element}
+  </Auth0Provider>
+)
```
Comment (jasonLaster) [inline site/gatsby-browser.js:20]: This is a great place to put this

File: site/src/components/Auth0Provider.js [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        isAuthenticated,
+        user,
+        loading,
+        popupOpen,
+        loginWithPopup,
+        handleRedirectCallback,
+        getIdTokenClaims: (...p) => auth0Client.getIdTokenClaims(...p),
+        loginWithRedirect: (...p) => auth0Client.loginWithRedirect(...p),
+        getTokenSilently: (...p) => auth0Client.getTokenSilently(...p),
+        getTokenWithPopup: (...p) => auth0Client.getTokenWithPopup(...p),
```
Comment (jasonLaster) [inline site/src/components/Auth0Provider.js:101]: Not important, but I wonder if you can do:

```js
getTokenWithPopup:  auth0Client.getTokenWithPopup
```

**Comment by jasonLaster** [comment]
2 other thoughts:

1. it'd be nice to show an ugly signout button in the header as a POC to show how the state can be shred
2. netlify build is failing, i bet we need some more env vars in netlify. or fix something for `gatsby build` in node

File: site/src/components/Auth0Provider.js (hunk 2 above)
Comment (bloudermilk) [reply site/src/components/Auth0Provider.js:101]: Yeah I was hoping to do the same. The trick is that `auth0Client` isn't loaded yet when that code runs, which is why it needs to be wrapped. FWIW most of this code is copy/pasted from the Auth0 docs.

**Comment by bloudermilk** [comment]
@jasonLaster regarding your points

1. I actually had a "pretty" button in the header but wanted to be able to integrate this into master w/o exposing it to the general public.
2. Working on that now. Typical Gatsby SSR issues :)

**Pull request merged**